src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,497 @@
+-- isSubgraphOf ('vertex' x)    'empty'         ==  False
+-- isSubgraphOf x             ('overlay' x y) ==  True
+-- isSubgraphOf ('overlay' x y) ('connect' x y) ==  True
+-- isSubgraphOf ('path' xs)     ('circuit' xs)  ==  True
+-- isSubgraphOf x y                         ==> x <= y
+-- @
+isSubgraphOf :: Ord a => SymmetricRelation a -> SymmetricRelation a -> Bool
+isSubgraphOf x y = R.isSubgraphOf (fromSymmetric x) (fromSymmetric y)
 
+-- | Check if a relation is empty.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- isEmpty 'empty'                       == True
+-- isEmpty ('overlay' 'empty' 'empty')       == True
+-- isEmpty ('vertex' x)                  == False
+-- isEmpty ('removeVertex' x $ 'vertex' x) == True
+-- isEmpty ('removeEdge' x y $ 'edge' x y) == False
+-- @
+isEmpty :: SymmetricRelation a -> Bool
+isEmpty = R.isEmpty . fromSymmetric
+
+-- | Check if a graph contains a given vertex.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasVertex x 'empty'            == False
+-- hasVertex x ('vertex' x)       == True
+-- hasVertex 1 ('vertex' 2)       == False
+-- hasVertex x . 'removeVertex' x == 'const' False
+-- @
+hasVertex :: Ord a => a -> SymmetricRelation a -> Bool
+hasVertex x = R.hasVertex x . fromSymmetric
+
+-- | Check if a graph contains a given edge.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasEdge x y 'empty'            == False
+-- hasEdge x y ('vertex' z)       == False
+-- hasEdge x y ('edge' x y)       == True
+-- hasEdge y x ('edge' x y)       == True
+-- hasEdge x y . 'removeEdge' x y == 'const' False
+-- hasEdge x y                  == 'elem' (x,y) . 'edgeList'
+-- @
+hasEdge :: Ord a => a -> a -> SymmetricRelation a -> Bool
+hasEdge x y = R.hasEdge x y . fromSymmetric
+
+-- | The number of vertices in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- vertexCount 'empty'             ==  0
+-- vertexCount ('vertex' x)        ==  1
+-- vertexCount                   ==  'length' . 'vertexList'
+-- vertexCount x \< vertexCount y ==> x \< y
+-- @
+vertexCount :: SymmetricRelation a -> Int
+vertexCount = R.vertexCount . fromSymmetric
+
+-- | The number of edges in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- edgeCount 'empty'      == 0
+-- edgeCount ('vertex' x) == 0
+-- edgeCount ('edge' x y) == 2
+-- edgeCount            == 'length' . 'edgeList'
+-- @
+edgeCount :: SymmetricRelation a -> Int
+edgeCount = R.edgeCount . fromSymmetric
+
+-- | The sorted list of vertices of a given graph.
+-- Complexity: /O(n)/ time and memory.
+--
+-- @
+-- vertexList 'empty'      == []
+-- vertexList ('vertex' x) == [x]
+-- vertexList . 'vertices' == 'Data.List.nub' . 'Data.List.sort'
+-- @
+vertexList :: SymmetricRelation a -> [a]
+vertexList = R.vertexList . fromSymmetric
+
+-- | The sorted list of edges of a graph.
+-- Complexity: /O(n + m)/ time and /O(m)/ memory.
+--
+-- @
+-- edgeList 'empty'          == []
+-- edgeList ('vertex' x)     == []
+-- edgeList ('edge' x y)     == [(x,y), (y,x)]
```

Comment:
Same here: semantically, `edge 1 2` and `edge 2 1` is the same edge, so you are listing the same edge twice.

Reply:
You should probably report edges in the order of their vertices, i.e. `edgeList (edge 1 2) = edgeList (edge 2 1) = [(1,2)]`.

Reply:
I fail to understand if you think it's best to store only 1 edge or not... I think it would be better to store the 2 edges so `edgeList (edge 1 2) = edgeList (edge 2 1) = [(1,2),(2,1)]` because that's the symmetric closure, but it's you're call! I might have to redo all of my work tho

Reply:
Storing both `(1,2)` and `(2,1)` is an implementation detail, it should not leak through the user interface. In undirected graphs edges are not pairs `(x,y)`, but sets `{x,y}`, and the order of elements in a set is irrelevant. Sets `{1,2}` and `{2,1}` are the same, so we shouldn't report both of them.

Reply:
I think I got it... can you please answer my last comment to confirm?